### ISSUE TITLE
Lane color and Lane type Segmentation

### ISSUE THREAD

**Issue by nannapaneni4** [datvuthanh/HybridNets#20]
In the traffic light and traffic sign detection issue, you've mentioned that we just have to change the obj_list in `project.yml` by adding the classes needed. Does that apply to seg_list as well? 

If I want to detect lane color and lane type, can I change `seg_list` as follows

`seg_list: ['road', 'double white', 'double yellow', 'single white', 'single yellow', 'solid', 'dashed']`

Actually, I need classes like `double solid yellow, double solid white, single solid yellow, single solid white, single dashed yellow, single dashed white, double dashed yellow, double dashed white`, but BDD100K already has labeled  `double white, double yellow, single white, single yellow` classes under Lane Categories and `solid, dashed` classes under Lane Styles

Will the change in `seg_list`  as shown above work, if not, how to do it

**Comment by xoiga123** [comment]
Yes and no.

Yes because the model won't break, segmentation output is softmax.
```python
self.segmentation_head = SegmentationHead(
            in_channels=64,
            out_channels=self.seg_classes+1 if self.seg_classes > 1 else self.seg_classes,
            activation='softmax2d' if self.seg_classes > 1 else 'sigmoid',
            kernel_size=1,
            upsampling=4,
        )
```

No because data preparation and demo breaks, we hardcoded in certain places.
```python
# projects.yml
dataset:
  color_rgb: false
  dataroot: ./datasets/bdd100k
  labelroot: ./datasets/data2/zwt/bdd/bdd100k/labels/100k
  laneroot: ./datasets/bdd_lane_gt  # this
  maskroot: ./datasets/bdd_seg_gt  # and this
```
```python
# dataset.py
_, seg1 = cv2.threshold(seg_label, 1, 255, cv2.THRESH_BINARY)
_, lane1 = cv2.threshold(lane_label, 1, 255, cv2.THRESH_BINARY)
# prefer lane
seg1 = seg1 - (seg1 & lane1)

union = seg1 | lane1
background = 255 - union
```
We'll softcode these parts _soon_, _hopefully_ by the next commit. In the meantime, please feel free to track where `seg_list` is used and modify them yourself if you fancy a challenge. Copy and paste lane to lane2...laneN is fine, but a few loops and logic alterations to generalize and we'll be more than happy to accept your wonderful pull request :wink:

**Comment by Laulian** [comment]
It seems doesn't work when I change the "seg_list" from "['road', 'lane']" to "['road', 'double white', 'double yellow', 'single white', 'single yellow', 'solid', 'dashed']". Should i do some other customized modify?

**Comment by xoiga123** [comment]
@Laulian we only provide a customized lane-line label (as stated in the paper), which doesn't contain all the classes you've listed. For the original lane-line label, please refer to [the original dataset](https://doc.bdd100k.com/download.html). After downloading, you have to convert [a single png with every class information](https://doc.bdd100k.com/format.html#lane-mask) into a png for each class that you want. You can utilize [the official BDD100K helper](https://github.com/bdd100k/bdd100k/tree/master/bdd100k) to do just that. When you're done, specify their corresponding paths like:

```yaml
seg_list: ['road', 'double white', 'double yellow', 'single white', 'single yellow', 'solid', 'dashed']

dataset:
  segroot:
  - path/to/road_pngs
  - path/to/double_white_pngs
  - path/to/double_yellow_pngs
  - ...
```

I'm sorry to say that this is dataset preparation and kinda goes beyond our scope. Try the helper repo, I believe people have successfully done it.

**Comment by Laulian** [comment]
@xoiga123  I see, thanks for the detailed reply, will try it. BTW, i have found a bug in your new val.py at line142,feel free.

**Comment by xoiga123** [comment]
@Laulian @nannapaneni4 do you guys have any suggestion for multi-label segmentation display when inferring? Normally, each class has their own color. Or should I use the order of classes in `seg_list` as their importance, whichever is first gets to have their color displayed?

**Comment by xoiga123** [comment]
Man, I've got most of the code done for multi-label (no pretrained of course, tested and ran fine for a few epochs), but what really got me stuck is inference display, how do you want the information conveniently delivered to you?

To be fair, this list is multi-class and each class can be assigned to a color:
```bash
seg_list = ['double solid yellow', 'double solid white', 'single solid yellow', 'single solid white', 'single dashed yellow', 'single dashed white', 'double dashed yellow', 'double dashed white']
```

While this list is multi-label, and is ass to learn (I guess) and display:
```bash
seg_list: ['road', 'double white', 'double yellow', 'single white', 'single yellow', 'solid', 'dashed']
```

**Comment by mickey-cool** [comment]
Excuse me,I found that when I finished the first epoch, the training process was killed automatically, do you know what is the problem?

**Comment by xoiga123** [comment]
@mickey-cool please check [this comment](https://github.com/datvuthanh/HybridNets/issues/16#issuecomment-1095370241).

**Comment by Yutong-gannis** [comment]
how can I get Lane type Segmentation dataset
